trac/web/main.py

```diff
@@ -883,14 +882,19 @@ def get_environments(environ, warn=False):
     env_parent_dir = environ.get('trac.env_parent_dir')
     if env_parent_dir:
         env_parent_dir = os.path.normpath(env_parent_dir)
-        paths = dircache.listdir(env_parent_dir)[:]
-        dircache.annotate(env_parent_dir, paths)
-
+        paths = [
+            name for name in os.listdir(env_parent_dir)
+            if os.path.isdir(os.path.join(env_parent_dir, name))
+        ]
```

Comment:
The changes are incorrect.

`dircache.annotate()` appends a `/` character to each path which points a directory in `paths`. However, the changes retrieve _only_ directories from `paths`.

See https://docs.python.org/2/library/dircache.html#dircache.annotate

Reply:
Doesn't the loop below only care about directories? (`if path[-1] == '/'`)

Reply:
Sorry. You're right. I've misunderstood.

I suggest the following changes:
```diff
diff --git a/trac/web/main.py b/trac/web/main.py
index 4cd5a255a..ce28ab680 100644
--- a/trac/web/main.py
+++ b/trac/web/main.py
@@ -883,14 +883,12 @@ def get_environments(environ, warn=False):
     env_parent_dir = environ.get('trac.env_parent_dir')
     if env_parent_dir:
         env_parent_dir = os.path.normpath(env_parent_dir)
-        paths = dircache.listdir(env_parent_dir)[:]
-        dircache.annotate(env_parent_dir, paths)
-
         # Filter paths that match the .tracignore patterns
         ignore_patterns = get_tracignore_patterns(env_parent_dir)
-        paths = [path[:-1] for path in paths if path[-1] == '/'
-                 and not any(fnmatch.fnmatch(path[:-1], pattern)
-                             for pattern in ignore_patterns)]
+        paths = [name for name in os.listdir(env_parent_dir)
+                      if os.path.isdir(os.path.join(env_parent_dir, name)) and
+                         not any(fnmatch.fnmatch(name, pattern)
+                                 for pattern in ignore_patterns)]
         env_paths.extend(os.path.join(env_parent_dir, project)
                          for project in paths)
     envs = {}
```

Reply:
Looks okay.